pkg/driver/node.go

```diff
@@ -71,30 +73,57 @@ func (d *Driver) NodeStageVolume(ctx context.Context, req *csi.NodeStageVolumeRe
 		return nil, status.Error(codes.InvalidArgument, "Volume capability not supported")
 	}
 
+	ok := d.inFlight.Insert(req)
+	if !ok {
+		msg := fmt.Sprintf("request to stage volume=%+v is already in process: formatting volume", volumeID)
+		return nil, status.Error(codes.Internal, msg)
+	}
+	defer func() {
+		klog.Infof("NodeStageVolume volume=%+v finished", req.GetVolumeId())
+		d.inFlight.Delete(req)
+	}()
+
 	source, ok := req.PublishContext["devicePath"]
 	if !ok {
 		return nil, status.Error(codes.InvalidArgument, "Device path not provided")
 	}
 
-	// TODO: consider replacing IsLikelyNotMountPoint by IsNotMountPoint
-	notMnt, err := d.mounter.Interface.IsLikelyNotMountPoint(target)
+	exists, err := d.mounter.Interface.ExistsPath(target)
 	if err != nil {
-		if os.IsNotExist(err) {
-			if errMkDir := d.mounter.Interface.MakeDir(target); errMkDir != nil {
-				msg := fmt.Sprintf("could not create target dir %q: %v", target, errMkDir)
-				return nil, status.Error(codes.Internal, msg)
-			}
-			notMnt = true
-		} else {
-			msg := fmt.Sprintf("could not determine if %q is valid mount point: %v", target, err)
+		msg := fmt.Sprintf("failed to check if target %+v exists: %+v", target, err)
+		return nil, status.Error(codes.Internal, msg)
+	}
+
+	// Check if a device is mounted in target directory
+	device, _, err := mount.GetDeviceNameFromMount(d.mounter, target)
+	if err != nil {
+		msg := fmt.Sprintf("failed to check if volume is already mounted: %+v", err)
+		return nil, status.Error(codes.Internal, msg)
+	}
+
+	// This operation (NodeStageVolume) MUST be idempotent. If the
+	// volume corresponding to the volume_id is already staged to
+	// the staging_target_path, and is identical to the specified
+	// volume_capability the Plugin MUST reply 0 OK.
+	if device == source {
+		klog.Infof("NodeStageVolume: volume=%+v already staged", volumeID)
+		return &csi.NodeStageVolumeResponse{}, nil
+	}
+
+	if exists {
```

Comment:
Having an empty `if` feels really strange, maybe `!exists` and then just put the comment below the `if`?